### PR TITLE
NIFI-10500: Improved property ordering in MQTT processors

### PR DESCRIPTION
**Pull request by turcsanyip**
# Summary

[NIFI-10500](https://issues.apache.org/jira/browse/NIFI-10500)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [ ] JDK 8
  - [ ] JDK 11
  - [ ] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/PublishMQTT.java [hunk 1]
```diff
@@ -117,23 +118,32 @@ public class PublishMQTT extends AbstractMQTTProcessor implements MqttCallback {
...
+            PROP_USERNAME,
+            PROP_PASSWORD,
+            PROP_SSL_CONTEXT_SERVICE,
+            PROP_CLEAN_SESSION,
+            PROP_SESSION_EXPIRY_INTERVAL,
+            PROP_CLIENTID,
+            PROP_TOPIC,
+            PROP_RETAIN,
+            PROP_QOS,
+            RECORD_READER,
```
Comment (nandorsoma) [inline nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/PublishMQTT.java:133]: I'd probably put reader/writer properties to the end of the list to keep the mqtt related properties together.

File: nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/common/AbstractMQTTProcessor.java [hunk 2]
```diff
@@ -307,7 +277,7 @@ public Collection<ValidationResult> customValidate(final ValidationContext valid
         final boolean readerIsSet = validationContext.getProperty(BASE_RECORD_READER).isSet();
         final boolean writerIsSet = validationContext.getProperty(BASE_RECORD_WRITER).isSet();
         if ((readerIsSet && !writerIsSet) || (!readerIsSet && writerIsSet)) {
-            results.add(new ValidationResult.Builder().subject("Reader and Writer").valid(false)
+            results.add(new ValidationResult.Builder().subject("Record Reader and Writer").valid(false)
                     .explanation("both Record Reader and Writer must be set when used.").build());
```
Comment (nandorsoma) [inline nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/common/AbstractMQTTProcessor.java:281]: Rendered text on the UI looks like "Record Reader and Writer is invalid, because both Record Reader and Writer must be set when used." which is a bit verbose. I'd remove the reference to "Record Reader and Writer" from the explanation to make the description simpler.

File: nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/ConsumeMQTT.java [hunk 3]
```diff
@@ -161,15 +162,26 @@ public class ConsumeMQTT extends AbstractMQTTProcessor implements MqttCallback {
...
+    public static final PropertyDescriptor RECORD_WRITER = new PropertyDescriptor.Builder()
+            .fromPropertyDescriptor(BASE_RECORD_WRITER)
+            .description("The Record Writer to use for serializing Records before writing them to a FlowFile.")
+            .build();
+
     public static final PropertyDescriptor ADD_ATTRIBUTES_AS_FIELDS = new PropertyDescriptor.Builder()
             .name("add-attributes-as-fields")
             .displayName("Add attributes as fields")
-            .description("If using the Records reader/writer and if setting this property to true, default fields "
+            .description("If using the Record Reader/Writer and if setting this property to true, default fields "
```
Comment (nandorsoma) [inline nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/ConsumeMQTT.java:178]: Because of the dependsOn property, the first if will always be true. I'd remove it to make the description simpler.

File: nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/ConsumeMQTT.java [hunk 4]
```diff
@@ -214,26 +216,35 @@ public class ConsumeMQTT extends AbstractMQTTProcessor implements MqttCallback {
...
+            PROP_USERNAME,
+            PROP_PASSWORD,
+            PROP_SSL_CONTEXT_SERVICE,
+            PROP_CLEAN_SESSION,
+            PROP_SESSION_EXPIRY_INTERVAL,
+            PROP_CLIENTID,
+            PROP_GROUPID,
+            PROP_TOPIC_FILTER,
+            PROP_QOS,
+            PROP_MAX_QUEUE_SIZE,
```
Comment (nandorsoma) [inline nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/ConsumeMQTT.java:231]: I'd put this property to the end of the list (before reader/writers if you accept that comment) because this property is not strictly related to the MQTT client itself as the other properties around it.

**Review by nandorsoma — CHANGES_REQUESTED** [review]
Thank you for your changes, @turcsanyip. I've found a few minor things that could be improved. Could you check them?

File: nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/PublishMQTT.java (hunk 1 above)
Comment (turcsanyip) [reply nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/PublishMQTT.java:133]: There are 2 main principles when ordering properties:
* properties with higher importance / being used frequently should come first
* keep related properties together

I believe the Timeout and Last Will properties are the least frequently used ones. Also, the by default visible Record properties may encourage their use. For these reasons I would keep this order even if those properties rather belong to the MQTT config. Record Writer has a similar position in the List* processors.

**Review by nandorsoma — APPROVED** [review]
Thanks @turcsanyip, LGTM!

**Review by arpadboda — APPROVED** [review]

**Comment by arpadboda** [comment]
Merged in https://github.com/apache/nifi/commit/5303aadda3e2c857dff4db8d20cdd4ec70722799
